src/coprocessor/dag/rpn_expr/impl_cast.rs

```diff
@@ -165,15 +169,15 @@ pub fn cast_uint_as_decimal(
 /// The signed int implementation for push down signature `CastIntAsDecimal`.
 #[rpn_fn(capture = [ctx, extra])]
 #[inline]
-pub fn cast_int_as_decimal(
+pub fn cast_any_as_decimal<From: Evaluable + ConvertTo<Decimal>>(
     ctx: &mut EvalContext,
     extra: &RpnFnCallExtra<'_>,
-    val: &Option<i64>,
+    val: &Option<From>,
 ) -> Result<Option<Decimal>> {
     match val {
         None => Ok(None),
         Some(val) => {
-            let dec = Decimal::from(*val);
+            let dec: Decimal = val.convert(ctx)?;
             Ok(Some(produce_dec_with_specified_tp(
```

Comment:
Looks like `produce_dec_with_specified_tp` should be placed inside the convert function, instead of `cast_xxx_as_xxx`. Otherwise we are still leaking cast logic in the non-cast layer, right?